perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -3090,6 +3120,15 @@ public List<RichGroup> getAllRichSubGroupsWithAttributesByNames(PerunSession ses
 		return convertGroupsToRichGroupsWithAttributes(sess, this.getAllSubGroups(sess, parentGroup), attrNames);
 	}
 
+	@Override
+	public List<RichGroup> getAllRichSubGroupsWithAttributesByNames(PerunSession sess, Group parentGroup, List<String> attrNames, List<String> roles, List<String> types) {
+		List<Group> subGroups = this.getAllSubGroups(sess, parentGroup);
+		if ((!roles.isEmpty() || !types.isEmpty()) && sess.getPerunPrincipal().getUser() != null) {
+			subGroups.removeIf(group -> !AuthzResolverBlImpl.groupMatchesUserRolesFilter(sess, sess.getPerunPrincipal().getUser(), group, roles, types));
```

Comment:
Calling this for every group doesn't seem very efficient. Why not pass roles and types to Impl layer and create a private method like `getSQLWhereForRolesAndTypes` which could be then reused for every modified method (you can extract the method already from `getGroupsPage`)
Just a thought, not really sure how much this actually effects speed